// material-community-components/esm2015/material-community-components.js
import { InjectionToken, Injectable, Inject, Component, ChangeDetectionStrategy, ChangeDetectorRef, EventEmitter, Input, Output, ContentChildren, ElementRef, Renderer2, ViewChild, Directive, forwardRef, NgModule } from '@angular/core';
import { BehaviorSubject, fromEvent } from 'rxjs';
import { coerceBooleanProperty } from '@angular/cdk/coercion';
import { map, debounceTime, withLatestFrom } from 'rxjs/operators';
import { FormBuilder, FormControl, Validators, NG_VALUE_ACCESSOR, ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';
import { OverlayModule } from '@angular/cdk/overlay';
import { PortalModule } from '@angular/cdk/portal';
import { MatButtonModule, MatFormFieldModule, MatInputModule } from '@angular/material';
import { ScrollDispatchModule } from '@angular/cdk/scrolling';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Contant used as empty color
 */
const EMPTY_COLOR = new InjectionToken('empty-color');
/**
 *
 * @record
 */

/**
 * This interface represents one color. Using this interface instead simple string
 * will help screen readers, because the text attribute ir set to the aria-label of
 * the option
 * @record
 */

/**
 * Verify if color has # as a first char. If not, add this char
 * to the color
 * @param {?} color string
 * @return {?}
 */
function coerceHexaColor(color) {
    if (color && color.indexOf('#') !== 0) {
        color = `#${color}`;
    }
    if (!isValidColor(color)) {
        return;
    }
    return color.toUpperCase();
}
/**
 * Validate if the color is valid
 * @param {?} color string
 * @return {?}
 */
function isValidColor(color) {
    // validate if color is an hexadecimal
    if (!color || color.charAt(0) !== '#' || color.length > 7) {
        return false;
    }
    // validate rgb of the color
    const /** @type {?} */ hex = color.replace('#', '');
    if (hex.length === 6) {
        const /** @type {?} */ r = parseInt(hex.slice(0, 2), 16);
        const /** @type {?} */ g = parseInt(hex.slice(2, 4), 16);
        const /** @type {?} */ b = parseInt(hex.slice(4, 6), 16);
        if (Number.isNaN(r) || Number.isNaN(g) || Number.isNaN(b)) {
            return false;
        }
    }
    return true;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccColorPickerService {
    /**
     * @param {?} emptyColor
     */
    constructor(emptyColor) {
        this.emptyColor = emptyColor;
        /**
         * Array of all used colors
         */
        this._colors = new BehaviorSubject([]);
    }
    /**
     * Add new color to used colors
     * @param {?} color string
     * @return {?}
     */
    addColor(color) {
        if (!color || !isValidColor(color)) {
            return;
        }
        color = coerceHexaColor(color) || this.emptyColor;
        const /** @type {?} */ colors = this._colors.getValue();
        if (!colors.find(_color => _color === color)) {
            colors.push(color);
            this._colors.next(colors);
        }
    }
    /**
     * Return Observable of colors
     * @return {?}
     */
    getColors() {
        return this._colors.asObservable();
    }
    /**
     * Reset the array of used colors
     * @return {?}
     */
    resetUseColors() {
        this._colors.next([]);
    }
}
MccColorPickerService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
MccColorPickerService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [EMPTY_COLOR,] },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccColorPickerCollectionComponent {
    /**
     * @param {?} changeDetectorRef
     * @param {?} emptyColor
     */
    constructor(changeDetectorRef, emptyColor) {
        this.changeDetectorRef = changeDetectorRef;
        this.emptyColor = emptyColor;
        this._hideEmpty = false;
        /**
         * Size limit of the collection
         */
        this.size = 30;
        /**
         * Show transparent option
         */
        this.transparent = false;
        /**
         * Emit selected color value
         */
        this.changeColor = new EventEmitter();
    }
    /**
     * Hide empty slots
     * Empty slots are the difference between the collection size and limit
     * @param {?} value
     * @return {?}
     */
    set hideEmpty(value) {
        this._hideEmpty = coerceBooleanProperty(value);
    }
    /**
     * Name of the collection
     * @return {?}
     */
    get label() {
        return this._label;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set label(value) {
        this._label = value;
    }
    /**
     * Array of colors to be displayed
     * @return {?}
     */
    get colors() {
        return this._colors;
    }
    /**
     * @param {?} values
     * @return {?}
     */
    set colors(values) {
        this._colors = values;
    }
    /**
     * @return {?}
     */
    ngAfterContentChecked() {
        if (this._colors && this._colors.length !== this.size) {
            this._colors = this._colors
                .slice(0, this.size)
                .concat(new Array(this._getCollectionDiffSize()));
            this.changeDetectorRef.markForCheck();
        }
    }
    /**
     * Return the difference between the limit and the collection size.
     * Always return 0 when hideEmpty is true
     * @return {?} number
     */
    _getCollectionDiffSize() {
        if (this._colors.length > this.size || this._hideEmpty) {
            return 0;
        }
        return this.size - this._colors.length;
    }
    /**
     * Remove color
     * @return {?}
     */
    setTransparent() {
        this.changeColor.emit(this.emptyColor);
    }
    /**
     * Emit selected color value
     * @param {?} option MccColorPickerOption
     * @return {?}
     */
    setColor(option) {
        const /** @type {?} */ color = typeof option === 'string' ? option : option.value;
        this.changeColor.emit(color);
    }
}
MccColorPickerCollectionComponent.decorators = [
    { type: Component, args: [{
                selector: 'mcc-color-picker-collection',
                template: `<div class="mcc-color-picker-collection" role="listbox" aria-label="Select a color">
        <h3>{{ label }}</h3>
        <button mat-mini-fab *ngIf="transparent" class="mcc-color-picker-remove-color mat-elevation-z0" role="option" aria-label="transparent"
                (click)="setTransparent()">
                <div class="mcc-color-picker-transparent"></div>
        </button>
        <button *ngFor="let color of colors" mat-mini-fab class="mat-elevation-z0" role="option" [disabled]="!color || color === emptyColor"
                [mccColorPickerOption]="color" (click)="setColor(color)">
        </button>
</div>`,
                styles: [`.mcc-color-picker-collection{min-height:50px;padding:10px 18px}.mcc-color-picker-collection h3{color:#100214;text-transform:uppercase;font-family:'Open Sans',sans-serif;font-size:12px;font-weight:700;margin:0 0 10px}.mcc-color-picker-remove-color{background-color:#fff;border:1px solid #e1e1e1}.mcc-color-picker-remove-color .mcc-color-picker-transparent{width:20px;height:1px;border-bottom:1px solid red;transform:translateY(-4px) translateX(0) rotate(45deg);-webkit-transform:translateY(-4px) translateX(0) rotate(45deg)}button{width:22px;height:22px;cursor:pointer;margin:3px}`],
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
            },] },
];
/** @nocollapse */
MccColorPickerCollectionComponent.ctorParameters = () => [
    { type: ChangeDetectorRef, },
    { type: undefined, decorators: [{ type: Inject, args: [EMPTY_COLOR,] },] },
];
MccColorPickerCollectionComponent.propDecorators = {
    "hideEmpty": [{ type: Input },],
    "label": [{ type: Input },],
    "colors": [{ type: Input },],
    "size": [{ type: Input },],
    "transparent": [{ type: Input },],
    "changeColor": [{ type: Output },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccColorPickerComponent {
    /**
     * @param {?} elementRef
     * @param {?} changeDetectorRef
     * @param {?} colorPickerService
     * @param {?} emptyColor
     */
    constructor(elementRef, changeDetectorRef, colorPickerService, emptyColor) {
        this.elementRef = elementRef;
        this.changeDetectorRef = changeDetectorRef;
        this.colorPickerService = colorPickerService;
        this.emptyColor = emptyColor;
        this._usedColorLabel = 'Used Colors';
        this._reverseUsedColor = false;
        this._hideHexForms = false;
        this._hideEmpty = false;
        this._hideTransparent = false;
        this._hideUsedColors = false;
        this._isOpen = false;
        this._overlay = true;
        this._hideButtons = false;
        this._colorPickerSelectorHeight = 170;
        this._hideColorPickerSelector = false;
        /**
         * Set the size of the used colors
         */
        this.usedSizeColors = 30;
        /**
         * Change btnCancel label
         */
        this.btnCancel = 'Cancel';
        /**
         * Change btnConfirm label
         */
        this.btnConfirm = 'Confirm';
        /**
         * Event emitted when user change the selected color (without confirm)
         */
        this.change = new EventEmitter();
        /**
         * Event emitted when selected color is confirm
         */
        this.selected = new EventEmitter();
        /**
         * Event emitted when is clicked outside of the component
         */
        this.clickOut = new EventEmitter();
        /**
         * Array of subscriptions from the collections
         */
        this._collectionSubs = [];
    }
    /**
     * Change label of the collection UsedColors
     * @return {?}
     */
    get usedColorLabel() {
        return this._usedColorLabel;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set usedColorLabel(value) {
        this._usedColorLabel = value;
    }
    /**
     * Set initial value for used color
     * @param {?} colors
     * @return {?}
     */
    set usedColorStart(colors) {
        if (colors && colors.length > 0) {
            for (const /** @type {?} */ color of colors) {
                this.colorPickerService.addColor(color);
            }
        }
    }
    /**
     * Set usedColor to be used in reverse
     * @param {?} reverse
     * @return {?}
     */
    set reverseUsedColors(reverse) {
        this._reverseUsedColor = coerceBooleanProperty(reverse);
    }
    /**
     * Hide the hexadecimal color forms.
     * @return {?}
     */
    get hideHexForms() {
        return this._hideHexForms;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set hideHexForms(value) {
        this._hideHexForms = value;
    }
    /**
     * Hide empty slots from the collection UsedColors
     * @return {?}
     */
    get hideEmpty() {
        return this._hideEmpty;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set hideEmpty(value) {
        this._hideEmpty = coerceBooleanProperty(value);
    }
    /**
     * Hide transparent option of UsedColors
     * @return {?}
     */
    get hideTransparent() {
        return this._hideTransparent;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set hideTransparent(value) {
        this._hideTransparent = coerceBooleanProperty(value);
    }
    /**
     * Hide UsedColors collection
     * @return {?}
     */
    get hideUsedColors() {
        return this._hideUsedColors;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set hideUsedColors(value) {
        this._hideUsedColors = coerceBooleanProperty(value);
    }
    /**
     * Start with a color selected
     * @return {?}
     */
    get selectedColor() {
        return this._selectedColor;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set selectedColor(value) {
        if (this._selectedColor !== value) {
            this.changeDetectorRef.markForCheck();
        }
        this._selectedColor = coerceHexaColor(value) || this.emptyColor;
    }
    /**
     * Define if the panel will be initiated open
     * @return {?}
     */
    get isOpen() {
        return this._isOpen;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set isOpen(value) {
        this._isOpen = coerceBooleanProperty(value);
    }
    /**
     * Define if the panel will show in overlay or not
     * @return {?}
     */
    get overlay() {
        return this._overlay;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set overlay(value) {
        this._overlay = coerceBooleanProperty(value);
    }
    /**
     * Hide the action buttons (cancel/confirm)
     * @return {?}
     */
    get hideButtons() {
        return this._hideButtons;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set hideButtons(value) {
        this._hideButtons = coerceBooleanProperty(value);
    }
    /**
     * Define new height for the selector
     * @return {?}
     */
    get colorPickerSelectorHeight() {
        return this._colorPickerSelectorHeight;
    }
    /**
     * @param {?} height
     * @return {?}
     */
    set colorPickerSelectorHeight(height) {
        this._colorPickerSelectorHeight = height;
    }
    /**
     * Hide the color picker selector
     * @return {?}
     */
    get hideColorPickerSelector() {
        return this._hideColorPickerSelector;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set hideColorPickerSelector(value) {
        this._hideColorPickerSelector = coerceBooleanProperty(value);
    }
    /**
     * Return a Observable with the color the user is picking
     * @return {?}
     */
    get tmpSelectedColor$() {
        return this._tmpSelectedColor.asObservable();
    }
    /**
     * Observable with all the colors used by the user
     * @return {?}
     */
    get usedColors$() {
        return this.colorPickerService
            .getColors()
            .pipe(map(colors => (!this._reverseUsedColor ? colors : [...colors].reverse())));
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        if (!this._selectedColor) {
            this._selectedColor = this.emptyColor;
        }
        this._tmpSelectedColor = new BehaviorSubject(this._selectedColor);
    }
    /**
     * Walk throw all collections and subcribe to changes
     * @return {?}
     */
    ngAfterContentInit() {
        if (this._collections) {
            this._collections.forEach((collection) => {
                const /** @type {?} */ subscription = collection.changeColor.subscribe(color => {
                    this.updateTmpSelectedColor(color);
                });
                this._collectionSubs.push(subscription);
            });
        }
    }
    /**
     * Destroy all subscriptions
     * @return {?}
     */
    ngOnDestroy() {
        if (this._collectionSubs) {
            this._collectionSubs.forEach((subscription) => {
                if (subscription && !subscription.closed) {
                    subscription.unsubscribe();
                }
            });
        }
    }
    /**
     * Update selected color and emit the change
     * @return {?}
     */
    _updateSelectedColor() {
        if (this._isOpen || !this.overlay) {
            const /** @type {?} */ tmpSelectedColor = this._tmpSelectedColor.getValue();
            if (this._selectedColor !== tmpSelectedColor) {
                this._selectedColor = tmpSelectedColor;
                this.selected.next(this._selectedColor);
            }
            else {
                this.selected.emit(this._selectedColor);
            }
        }
    }
    /**
     * Open/close color picker panel
     * @return {?}
     */
    toggle() {
        this._isOpen = !this._isOpen;
        if (!this._isOpen && this._selectedColor !== this.emptyColor) {
            this.colorPickerService.addColor(this._selectedColor);
        }
    }
    /**
     * Update selected color, close the panel and notify the user
     * @return {?}
     */
    backdropClick() {
        if (this._hideButtons) {
            this.confirmSelectedColor();
        }
        else {
            this.cancelSelection();
        }
        this.clickOut.emit(null);
    }
    /**
     * Update tmpSelectedColor
     * @param {?} color string
     * @return {?}
     */
    updateTmpSelectedColor(color) {
        if (color) {
            this._tmpSelectedColor.next(color);
            this.change.next(color);
            if (this._hideButtons) {
                this._updateSelectedColor();
            }
        }
    }
    /**
     * Cancel the selection and close the panel
     * @return {?}
     */
    cancelSelection() {
        this.selected.emit(this._selectedColor);
        this.change.next(this._selectedColor);
        this.toggle();
    }
    /**
     * Update selectedColor and close the panel
     * @return {?}
     */
    confirmSelectedColor() {
        this._updateSelectedColor();
        this.toggle();
    }
}
MccColorPickerComponent.decorators = [
    { type: Component, args: [{
                selector: 'mcc-color-picker',
                template: `<!-- color picker overlay -->
<ng-container *ngIf="overlay">
    <button type="button" class="btn-picker" cdkOverlayOrigin #trigger="cdkOverlayOrigin" [ngClass]="{ 'empty': selectedColor === emptyColor }"
        [style.background]="selectedColor" (click)="toggle()">
        <div class="transparent" *ngIf="selectedColor === emptyColor"></div>
    </button>
    <ng-template cdkConnectedOverlay cdkConnectedOverlayHasBackdrop cdkConnectedOverlayBackdropClass="mcc-color-picker-backdrop"
        [cdkConnectedOverlayOrigin]="trigger" [cdkConnectedOverlayOpen]="isOpen" (backdropClick)="backdropClick()">
        <ng-template [cdkPortalOutlet]="overlayPanel"></ng-template>
    </ng-template>
</ng-container>
<!-- color picker flat -->
<ng-template *ngIf="!overlay" [cdkPortalOutlet]="overlayPanel"></ng-template>
<!-- color picker component content -->
<ng-template cdkPortal #overlayPanel="cdkPortal">
    <div class="mcc-color-picker-overlay mat-elevation-z6" role="dialog" aria-label="Color picker">
        <mcc-color-picker-selector *ngIf="!hideColorPickerSelector" [selectedColor]="tmpSelectedColor$ | async" [hideHexForms]="hideHexForms"
            [height]="colorPickerSelectorHeight" (changeSelectedColor)="updateTmpSelectedColor($event)">
        </mcc-color-picker-selector>
        <mcc-color-picker-collection *ngIf="!hideUsedColors" [label]="usedColorLabel" [size]="usedSizeColors" [transparent]="!hideTransparent"
            [hideEmpty]="hideEmpty" [colors]="usedColors$ | async" (changeColor)="updateTmpSelectedColor($event)">
        </mcc-color-picker-collection>
        <ng-content></ng-content>
        <div *ngIf="!hideButtons" class="mcc-color-picker-actions">
            <button mat-button role="button" aria-label="Cancel" (click)="cancelSelection()">
                {{ btnCancel }}
            </button>
            <button mat-button role="button" aria-label="Confirm" (click)="confirmSelectedColor()">
                {{ btnConfirm }}
            </button>
        </div>
    </div>
</ng-template>`,
                styles: [`@import url(https://fonts.googleapis.com/css?family=Open+Sans:400,700);.btn-picker{width:25px;height:25px;cursor:pointer;background:0 0;border:2px solid #ddd}.btn-picker.empty{background:#fff!important}.mcc-color-picker-overlay{display:-webkit-box;display:-ms-flexbox;display:flex;width:260px;min-height:80px;position:relative;-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;padding:0;background:#fff;font-family:'Open Sans',sans-serif}.mcc-color-picker-overlay .mcc-color-picker-preview{width:100%;height:8px}.transparent{width:32px;height:2px;border-bottom:2px solid red;transform:translateY(-3px) translateX(-2px) rotate(45deg);-webkit-transform:translateY(-2px) translateX(-11px) rotate(45deg);position:absolute}.mcc-color-picker-actions{display:-webkit-box;display:-ms-flexbox;display:flex;padding:4px;border-top:1px solid #ddd}.mcc-color-picker-actions button{color:#100214;text-transform:uppercase;font-family:'Open Sans',sans-serif;font-size:12px;font-weight:400;-webkit-box-flex:1;-ms-flex-positive:1;flex-grow:1}`],
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
            },] },
];
/** @nocollapse */
MccColorPickerComponent.ctorParameters = () => [
    { type: ElementRef, },
    { type: ChangeDetectorRef, },
    { type: MccColorPickerService, },
    { type: undefined, decorators: [{ type: Inject, args: [EMPTY_COLOR,] },] },
];
MccColorPickerComponent.propDecorators = {
    "_collections": [{ type: ContentChildren, args: [MccColorPickerCollectionComponent,] },],
    "usedColorLabel": [{ type: Input },],
    "usedColorStart": [{ type: Input },],
    "reverseUsedColors": [{ type: Input },],
    "hideHexForms": [{ type: Input, args: ['hideHexForms',] },],
    "hideEmpty": [{ type: Input, args: ['hideEmptyUsedColors',] },],
    "hideTransparent": [{ type: Input, args: ['hideTransparentUsedColors',] },],
    "hideUsedColors": [{ type: Input, args: ['hideUsedColors',] },],
    "selectedColor": [{ type: Input },],
    "isOpen": [{ type: Input },],
    "overlay": [{ type: Input },],
    "hideButtons": [{ type: Input },],
    "colorPickerSelectorHeight": [{ type: Input },],
    "hideColorPickerSelector": [{ type: Input },],
    "usedSizeColors": [{ type: Input },],
    "btnCancel": [{ type: Input },],
    "btnConfirm": [{ type: Input },],
    "change": [{ type: Output },],
    "selected": [{ type: Output },],
    "clickOut": [{ type: Output },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccColorPickerSelectorComponent {
    /**
     * @param {?} formBuilder
     * @param {?} render
     * @param {?} emptyColor
     */
    constructor(formBuilder, render, emptyColor) {
        this.formBuilder = formBuilder;
        this.render = render;
        this.emptyColor = emptyColor;
        this._height = 170;
        this._selectedColor = '';
        this._hideHexForms = false;
        /**
         * Emit update when a color is selected
         */
        this.changeSelectedColor = new EventEmitter();
        /**
         * RGBA current color
         */
        this._rgbaColor = 'rgba(255,0,0,1)';
        /**
         * Handle color of the text
         */
        this.textClass = 'black';
        /**
         * Validate if the mouse button is pressed
         */
        this._isPressed = false;
        /**
         * Form and keys of the fields in RGB
         */
        this.rgbKeys = ['R', 'G', 'B'];
    }
    /**
     * Canvas of the block
     * @param {?} el
     * @return {?}
     */
    set blockCursor(el) {
        this._bc = el;
    }
    /**
     * Container of the strip
     * @param {?} el
     * @return {?}
     */
    set stripCursor(el) {
        this._sc = el;
    }
    /**
     * Change height base of the selector
     * @param {?} value
     * @return {?}
     */
    set height(value) {
        this._height = value;
    }
    /**
     * @return {?}
     */
    get selectorHeight() {
        return this._height;
    }
    /**
     * @return {?}
     */
    get stripHeight() {
        return this._height - 10;
    }
    /**
     * Receive selected color from the component
     * @return {?}
     */
    get selectedColor() {
        return this._selectedColor;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set selectedColor(value) {
        this._selectedColor = value || this.emptyColor;
    }
    /**
     * Hide the hexadecimal color forms.
     * @return {?}
     */
    get hideHexForms() {
        return this._hideHexForms;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set hideHexForms(value) {
        this._hideHexForms = value;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this._tmpSelectedColor = new BehaviorSubject(this._selectedColor);
        this._tmpSelectedColorSub = this._tmpSelectedColor.subscribe(color => {
            if (color !== this._selectedColor && isValidColor(color)) {
                if (this.hexForm.get('hexCode').value !== color) {
                    this.hexForm.setValue({ hexCode: color });
                }
                this.changeSelectedColor.emit(coerceHexaColor(color) || this.emptyColor);
            }
        });
        // hex form
        this.hexForm = this.formBuilder.group({
            hexCode: [this.selectedColor, [Validators.minLength(7), Validators.maxLength(7)]],
        });
        // rgb dynamic form
        const /** @type {?} */ rgbGroup = {};
        const /** @type {?} */ rgbValue = this._getRGB();
        this.rgbKeys.forEach((key, index) => (rgbGroup[key] = new FormControl(rgbValue[index], {
            validators: [
                Validators.min(0),
                Validators.max(256),
                Validators.minLength(1),
                Validators.maxLength(3),
            ],
            updateOn: 'blur',
        })));
        this.rgbForm = this.formBuilder.group(rgbGroup);
        // watch changes on forms
        this._onChanges();
    }
    /**
     * Update RGB, RGBA and Gradient when selectedColor change and
     * the mouse button is pressed
     * @param {?} changes SimpleChanges
     * @return {?}
     */
    ngOnChanges(changes) {
        if ('selectedColor' in changes && changes['selectedColor'].currentValue !== this.emptyColor) {
            if (!this._isPressed) {
                this._updateRGB();
                this._updateRGBA();
                if (this._blockContext) {
                    this._fillGradient();
                }
            }
            const /** @type {?} */ rgb = this._getRGB();
            const /** @type {?} */ o = Math.round((rgb[0] * 299 + rgb[1] * 587 + rgb[2] * 114) / 1000);
            this.textClass = o > 125 ? 'black' : 'white';
        }
    }
    /**
     * Destroy all subscriptions
     * @return {?}
     */
    ngOnDestroy() {
        if (this._tmpSelectedColorSub && !this._tmpSelectedColorSub.closed) {
            this._tmpSelectedColorSub.unsubscribe();
        }
        if (this._hexValuesSub && !this._hexValuesSub.closed) {
            this._hexValuesSub.unsubscribe();
        }
        if (this._rgbValuesSub && !this._rgbValuesSub.closed) {
            this._rgbValuesSub.unsubscribe();
        }
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.render.listen(this._block.nativeElement, 'mousedown', e => {
            this._isPressed = true;
            this.changeColor(e);
        });
        this.render.listen(this._block.nativeElement, 'mouseup', () => (this._isPressed = false));
        this.render.listen(this._block.nativeElement, 'mouseout', () => (this._isPressed = false));
        this.render.listen(this._block.nativeElement, 'mousemove', e => this.changeColor(e));
        this._blockContext = this._bc.nativeElement.getContext('2d');
        this._blockContext.rect(0, 0, this._bc.nativeElement.width, this._bc.nativeElement.height);
        this.render.listen(this._strip.nativeElement, 'mousedown', e => {
            this._isPressed = true;
            this.changeBaseColor(e);
        });
        this.render.listen(this._strip.nativeElement, 'mouseup', () => (this._isPressed = false));
        this.render.listen(this._strip.nativeElement, 'mouseout', () => (this._isPressed = false));
        this.render.listen(this._strip.nativeElement, 'mousemove', e => this.changeBaseColor(e));
        this._stripContext = this._strip.nativeElement.getContext('2d');
        this._stripContext.rect(0, 0, this._strip.nativeElement.width, this._strip.nativeElement.height);
        const /** @type {?} */ grd1 = this._stripContext.createLinearGradient(0, 0, 0, this._bc.nativeElement.height);
        grd1.addColorStop(0, 'rgba(255, 0, 0, 1)');
        grd1.addColorStop(0.17, 'rgba(255, 255, 0, 1)');
        grd1.addColorStop(0.34, 'rgba(0, 255, 0, 1)');
        grd1.addColorStop(0.51, 'rgba(0, 255, 255, 1)');
        grd1.addColorStop(0.68, 'rgba(0, 0, 255, 1)');
        grd1.addColorStop(0.85, 'rgba(255, 0, 255, 1)');
        grd1.addColorStop(1, 'rgba(255, 0, 0, 1)');
        this._stripContext.fillStyle = grd1;
        this._stripContext.fill();
        this._fillGradient();
    }
    /**
     * Generate colors based on the RGBA color
     * @return {?}
     */
    _fillGradient() {
        this._blockContext.fillStyle = this._rgbaColor;
        this._blockContext.fillRect(0, 0, this._bc.nativeElement.width, this._bc.nativeElement.height);
        const /** @type {?} */ grdWhite = this._stripContext.createLinearGradient(0, 0, this._bc.nativeElement.width, 0);
        grdWhite.addColorStop(0, 'rgba(255,255,255,1)');
        grdWhite.addColorStop(1, 'rgba(255,255,255,0)');
        this._blockContext.fillStyle = grdWhite;
        this._blockContext.fillRect(0, 0, this._bc.nativeElement.width, this._bc.nativeElement.height);
        const /** @type {?} */ grdBlack = this._stripContext.createLinearGradient(0, 0, 0, this._bc.nativeElement.height);
        grdBlack.addColorStop(0, 'rgba(0,0,0,0)');
        grdBlack.addColorStop(1, 'rgba(0,0,0,1)');
        this._blockContext.fillStyle = grdBlack;
        this._blockContext.fillRect(0, 0, this._bc.nativeElement.width, this._bc.nativeElement.height);
    }
    /**
     * Watch change on forms
     * @return {?}
     */
    _onChanges() {
        // validate digited code and update when digitation is finished
        this._hexValuesSub = this.hexForm.get('hexCode').valueChanges.subscribe(value => {
            if (!this._isPressed && isValidColor(value)) {
                this._tmpSelectedColor.next(coerceHexaColor(value) || this.emptyColor);
            }
        });
        this._rgbValuesSub = this.rgbForm.valueChanges.subscribe(controls => {
            const /** @type {?} */ data = [];
            for (const /** @type {?} */ key in controls) {
                if (!controls[key] || controls[key] > 255) {
                    data.push('');
                    continue;
                }
                data.push(controls[key]);
            }
            const /** @type {?} */ hex = this._getHex(data);
            if (hex !== this._selectedColor && hex.length === 7) {
                this._tmpSelectedColor.next(hex);
            }
        });
    }
    /**
     * Convert HEX/canvas value to rgb
     * @param {?=} data any
     * @return {?} number[]
     */
    _getRGB(data) {
        if (data) {
            return [data[0], data[1], data[2]];
        }
        const /** @type {?} */ hex = this._selectedColor.replace('#', '');
        const /** @type {?} */ r = parseInt(hex.slice(0, 2), 16);
        const /** @type {?} */ g = parseInt(hex.slice(2, 4), 16);
        const /** @type {?} */ b = parseInt(hex.slice(4, 6), 16);
        return [r, g, b];
    }
    /**
     * Convert RGB value to HEX
     * @param {?} data any
     * @return {?} string
     */
    _getHex(data) {
        const /** @type {?} */ hex = new Array(3);
        hex[0] = data[0].toString(16);
        hex[1] = data[1].toString(16);
        hex[2] = data[2].toString(16);
        hex.forEach((val, key) => {
            if (val.length === 1) {
                hex[key] = '0' + hex[key];
            }
        });
        return coerceHexaColor(`${hex[0]}${hex[1]}${hex[2]}`) || this.emptyColor;
    }
    /**
     * Update RGBA color
     * @param {?=} data any
     * @return {?}
     */
    _updateRGBA(data) {
        if (!this._selectedColor && !data) {
            this._rgbaColor = 'rgba(255,0,0,1)';
        }
        const /** @type {?} */ rgb = this._getRGB(data);
        this._rgbaColor = `rgba(${rgb[0]}, ${rgb[1]}, ${rgb[2]}, 1)`;
    }
    /**
     * Update RGB form
     * @param {?=} data any
     * @return {?}
     */
    _updateRGB(data) {
        if (!this.rgbForm) {
            return;
        }
        if (!data) {
            data = this._getRGB();
        }
        this.rgbForm.setValue({ R: data[0], G: data[1], B: data[2] });
    }
    /**
     * Get selected base color from the canvas
     * @param {?} e MouseEvent
     * @return {?}
     */
    changeBaseColor(e) {
        if (this._isPressed) {
            this.render.setStyle(this._sc.nativeElement, 'background-position-y', `${e.offsetY}px`);
            const /** @type {?} */ data = this._stripContext.getImageData(e.offsetX, e.offsetY, 1, 1).data;
            this._updateRGBA(data);
            this._fillGradient();
            this.updateValues(data);
        }
    }
    /**
     * Get selected color from the canvas
     * @param {?} e MouseEvent
     * @return {?}
     */
    changeColor(e) {
        if (this._isPressed) {
            this.render.setStyle(this._bp.nativeElement, 'top', `${e.offsetY - 5}px`);
            this.render.setStyle(this._bp.nativeElement, 'left', `${e.offsetX - 5}px`);
            const /** @type {?} */ data = this._blockContext.getImageData(e.offsetX, e.offsetY, 1, 1).data;
            this.updateValues(data);
        }
    }
    /**
     * Emit update from the selected color
     * @param {?} data any
     * @return {?}
     */
    updateValues(data) {
        if (data) {
            this._updateRGB(data);
            this._tmpSelectedColor.next(this._getHex(data));
        }
    }
}
MccColorPickerSelectorComponent.decorators = [
    { type: Component, args: [{
                selector: 'mcc-color-picker-selector',
                template: `<div class="mcc-color-picker-selector" [ngStyle]="{ 'height.px': selectorHeight }">
    <div #block class="mcc-picker-selector"></div>
    <canvas #blockCanvas [height]="selectorHeight" width="230" id="picker"></canvas>
    <div #blockPointer class="mcc-picker-position" style="top: 0px;left: 220px;"></div>
    <div #stripContainer class="mcc-colors-position" style="background-position-y: 0px;">
        <canvas #strip [height]="stripHeight" width="20" id="colors"></canvas>
    </div>
</div>
<div class="mcc-color-picker-selector-preview" [style.background]="selectedColor">
    <ng-container *ngIf="!hideHexForms">
        <form [formGroup]="hexForm">
            <mat-form-field class="hex-input" floatLabel="always" [ngClass]="textClass">
                <input matInput placeholder="HEX" maxlength="7" formControlName="hexCode" [value]="selectedColor" />
            </mat-form-field>
        </form>
        <form [formGroup]="rgbForm">
            <mat-form-field class="rgb-input" floatLabel="always" [ngClass]="textClass">
                <input matInput type="number" placeholder="RGB" maxlength="3" formControlName="R" />
            </mat-form-field>
            <mat-form-field class="rgb-input" floatLabel="always" [ngClass]="textClass">
                <input matInput type="number" maxlength="3" formControlName="G" />
            </mat-form-field>
            <mat-form-field class="rgb-input" floatLabel="always" [ngClass]="textClass">
                <input matInput type="number" maxlength="3" formControlName="B" />
            </mat-form-field>
        </form>
    </ng-container>
</div>`,
                styles: [`canvas#colors{margin:5px}canvas:hover{cursor:crosshair}.mcc-picker-selector{position:absolute;width:230px;height:170px;top:0;left:0;z-index:10;cursor:crosshair}.mcc-picker-position{position:absolute;width:10px;height:10px;z-index:1;border:1px solid #ddd;border-radius:50%;background:rgba(0,0,0,.3)}.mcc-colors-position{position:absolute;width:30px;height:160px;top:0;right:0;z-index:1;background:url(data:image/gif;base64,R0lGODdhHgAIALMAAAAAADU1NTk5OUJCQkpKSlZWVltbW2pqaoCAgP///wAAAAAAAAAAAAAAAAAAAAAAACH5BAkAAAoALAAAAAAeAAgAAAQw0BhFq734yjJm/p8xFEVAeGCqGERACmahgkUrvPH81cFdnjoQy8VBBTOiguSokkQAADs=) no-repeat}.mcc-color-picker-selector{height:170px}/deep/ mat-form-field{font-family:'Open Sans';font-size:11px;font-weight:700}/deep/ mat-form-field.black{color:#100214}/deep/ mat-form-field.black label{color:#100214}/deep/ mat-form-field.black .mat-form-field-underline{background-color:#100214}/deep/ mat-form-field.white{color:#fff}/deep/ mat-form-field.white label{color:#fff}/deep/ mat-form-field.white .mat-form-field-underline{background-color:#fff}.mcc-color-picker-selector-preview{height:40px;padding:15px;border-bottom:1px solid #e1e1e1}.mcc-color-picker-selector-preview form{display:inline-block;position:relative}.mcc-color-picker-selector-preview .hex-input{width:75px;margin-right:20px}.mcc-color-picker-selector-preview .hex-input input{font-size:16px;font-weight:400}.mcc-color-picker-selector-preview .rgb-input{width:40px;margin-right:5px}.mcc-color-picker-selector-preview .rgb-input input{font-size:16px;font-weight:400}`],
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
            },] },
];
/** @nocollapse */
MccColorPickerSelectorComponent.ctorParameters = () => [
    { type: FormBuilder, },
    { type: Renderer2, },
    { type: undefined, decorators: [{ type: Inject, args: [EMPTY_COLOR,] },] },
];
MccColorPickerSelectorComponent.propDecorators = {
    "_block": [{ type: ViewChild, args: ['block',] },],
    "_bp": [{ type: ViewChild, args: ['blockPointer',] },],
    "blockCursor": [{ type: ViewChild, args: ['blockCanvas',] },],
    "_strip": [{ type: ViewChild, args: ['strip',] },],
    "stripCursor": [{ type: ViewChild, args: ['stripContainer',] },],
    "height": [{ type: Input, args: ['height',] },],
    "selectedColor": [{ type: Input },],
    "hideHexForms": [{ type: Input, args: ['hideHexForms',] },],
    "changeSelectedColor": [{ type: Output },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * This directive change the background of the button
 */
class MccColorPickerOptionDirective {
    /**
     * @param {?} elementRef
     * @param {?} render
     * @param {?} emptyColor
     */
    constructor(elementRef, render, emptyColor) {
        this.elementRef = elementRef;
        this.render = render;
        this.emptyColor = emptyColor;
        this._color = emptyColor;
    }
    /**
     * Receive the color
     * @return {?}
     */
    get color() {
        return this._color;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set color(value) {
        this._color = value;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.color) {
            let /** @type {?} */ color;
            if (typeof this.color === 'string') {
                color = this.color;
            }
            else {
                color = this.color.value;
                this.render.setAttribute(this.elementRef.nativeElement, 'aria-label', this.color.text);
            }
            if (isValidColor(color)) {
                // apply the color
                this.render.setStyle(this.elementRef.nativeElement, 'background', coerceHexaColor(color) || this.emptyColor);
            }
        }
    }
}
MccColorPickerOptionDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mccColorPickerOption], [mcc-color-picker-option]',
                exportAs: 'mccColorPickerOption',
            },] },
];
/** @nocollapse */
MccColorPickerOptionDirective.ctorParameters = () => [
    { type: ElementRef, },
    { type: Renderer2, },
    { type: undefined, decorators: [{ type: Inject, args: [EMPTY_COLOR,] },] },
];
MccColorPickerOptionDirective.propDecorators = {
    "color": [{ type: Input, args: ['mccColorPickerOption',] },],
};
/**
 * Directive applied to an element to make it usable as an origin for an ColorPicker.
 */
class MccColorPickerOriginDirective {
    /**
     * Reference to the element on which the directive is applied.
     * @param {?} elementRef
     * @param {?} renderer
     * @param {?} emptyColor
     */
    constructor(elementRef, renderer, emptyColor) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        this.emptyColor = emptyColor;
        /**
         * Emit changes from the origin
         */
        this.change = new BehaviorSubject('');
        // listen changes onkeyup and update color picker
        renderer.listen(elementRef.nativeElement, 'keyup', (event) => {
            const /** @type {?} */ value = event.currentTarget['value'];
            if (event.isTrusted && isValidColor(value)) {
                this.writeValueFromKeyup(coerceHexaColor(value) || this.emptyColor);
            }
        });
    }
    /**
     * This method will be called by the forms API to write to the view when
     * programmatic (model -> view) changes are requested.
     * @param {?} color
     * @return {?}
     */
    writeValue(color) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'value', color);
        this.change.next(color);
        if (this.propagateChanges) {
            this.propagateChanges(color);
        }
    }
    /**
     * This method will be called by the color picker
     * @param {?} color
     * @return {?}
     */
    writeValueFromColorPicker(color) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'value', color);
        this.propagateChanges(color);
    }
    /**
     * This method will be called from origin whe key is up
     * @param {?} color
     * @return {?}
     */
    writeValueFromKeyup(color) {
        this.change.next(color);
        this.propagateChanges(color);
    }
    /**
     * This is called by the forms API on initialization so it can update the
     * form model when values propagate from the view (view -> model).
     * @param {?} fn any
     * @return {?}
     */
    registerOnChange(fn) {
        this.propagateChanges = fn;
    }
    /**
     * This is called by the forms API on initialization so it can update the form model on blur
     * @param {?} fn any
     * @return {?}
     */
    registerOnTouched(fn) { }
    /**
     * called by the forms API when the control status changes to or from "DISABLED"
     * @param {?} isDisabled boolean
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
    }
}
MccColorPickerOriginDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mcc-color-picker-origin], [mccColorPickerOrigin]',
                exportAs: 'mccColorPickerOrigin',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => MccColorPickerOriginDirective),
                        multi: true,
                    },
                ],
            },] },
];
/** @nocollapse */
MccColorPickerOriginDirective.ctorParameters = () => [
    { type: ElementRef, },
    { type: Renderer2, },
    { type: undefined, decorators: [{ type: Inject, args: [EMPTY_COLOR,] },] },
];
MccColorPickerOriginDirective.propDecorators = {
    "change": [{ type: Output },],
};
/**
 * Directive connect an color picker with any input, select or textarea.
 * The color picker will be automatically updated when the value of the origin is
 * changed.
 */
class MccConnectedColorPickerDirective {
    /**
     * @param {?} colorPicker
     * @param {?} changeDetectorRef
     * @param {?} emptyColor
     */
    constructor(colorPicker, changeDetectorRef, emptyColor) {
        this.colorPicker = colorPicker;
        this.changeDetectorRef = changeDetectorRef;
        this.emptyColor = emptyColor;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (!this._colorPickerSub) {
            this._attachColorPicker();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._colorPickerSub && !this._colorPickerSub.closed) {
            this._colorPickerSub.unsubscribe();
        }
        if (this._originSub && !this._originSub.closed) {
            this._originSub.unsubscribe();
        }
    }
    /**
     * Attach color picker and origin
     * @return {?}
     */
    _attachColorPicker() {
        // subscribe to origin change to update color picker
        this._originSub = this.origin.change.subscribe(value => {
            if (isValidColor(value) ||
                (value === this.emptyColor && this.colorPicker.selectedColor !== this.emptyColor)) {
                this.colorPicker.updateTmpSelectedColor(value);
            }
            this.colorPicker.selectedColor = value;
            this.changeDetectorRef.detectChanges();
        });
        // subscribe to color picker changes and set on origin element
        this._colorPickerSub = this.colorPicker.change.subscribe(value => this.origin.writeValueFromColorPicker(value));
    }
}
MccConnectedColorPickerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mcc-connected-color-picker], [mccConnectedColorPicker]',
                exportAs: 'mccConnectedColorPicker',
            },] },
];
/** @nocollapse */
MccConnectedColorPickerDirective.ctorParameters = () => [
    { type: MccColorPickerComponent, },
    { type: ChangeDetectorRef, },
    { type: undefined, decorators: [{ type: Inject, args: [EMPTY_COLOR,] },] },
];
MccConnectedColorPickerDirective.propDecorators = {
    "origin": [{ type: Input, args: ['mccConnectedColorPickerOrigin',] },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccColorPickerModule {
    /**
     *
     * @param {?} config
     * @return {?}
     */
    static forRoot(config) {
        return {
            ngModule: MccColorPickerModule,
            providers: [{ provide: EMPTY_COLOR, useValue: config.empty_color }],
        };
    }
}
MccColorPickerModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    PortalModule,
                    OverlayModule,
                    ReactiveFormsModule,
                    MatButtonModule,
                    MatFormFieldModule,
                    MatInputModule,
                ],
                declarations: [
                    MccColorPickerComponent,
                    MccConnectedColorPickerDirective,
                    MccColorPickerSelectorComponent,
                    MccColorPickerOriginDirective,
                    MccColorPickerOptionDirective,
                    MccColorPickerCollectionComponent,
                ],
                exports: [
                    MccColorPickerComponent,
                    MccConnectedColorPickerDirective,
                    MccColorPickerOriginDirective,
                    MccColorPickerCollectionComponent,
                ],
                providers: [MccColorPickerService, { provide: EMPTY_COLOR, useValue: 'none' }],
            },] },
];
/** @nocollapse */
MccColorPickerModule.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * contants to create timer with HOURS or MINUTES
 */
const HOURS = ['12', '11', '1', '10', '2', '9', '3', '8', '4', '7', '5', '6'];
const MINUTES = [
    '00',
    '55',
    '05',
    '50',
    '10',
    '45',
    '15',
    '40',
    '20',
    '35',
    '25',
    '30',
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccTimerPickerComponent {
    constructor() {
        this._clock = new BehaviorSubject(HOURS);
        this._focus = 'hour';
        this._hour = '12';
        this._minute = '00';
        this._period = 'am';
        this._hideButtons = false;
        /**
         * Format of the hour to be emited on confirm
         */
        this.format = '12';
        /**
         * Change btnCancel label
         */
        this.btnCancel = 'Cancel';
        /**
         * Change btnConfirm label
         */
        this.btnConfirm = 'Ok';
        /**
         * Event emited when confirm button is pressed.
         * If buttons are hidden, the event is emited when value is changed
         */
        this.selected = new EventEmitter();
        /**
         * Set to true when timer picker have been connected with another component
         */
        this.connected = false;
    }
    /**
     * Current value (hour/minute) to create the clock
     * @return {?}
     */
    get clock$() {
        return this._clock.asObservable();
    }
    /**
     * Type there is in focus (hour/minute)
     * @return {?}
     */
    get focus() {
        return this._focus;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set focus(value) {
        if (value !== this._focus) {
            this._focus = value;
            this._clock.next(this._focus === 'hour' ? HOURS : MINUTES);
        }
    }
    /**
     * State of the overlay
     * @return {?}
     */
    get isOpen() {
        return this._isOpen;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set isOpen(value) {
        this._isOpen = coerceBooleanProperty(value);
    }
    /**
     * Return temporary selected hour (const HOURS)
     * @return {?}
     */
    get hour() {
        return this._hour;
    }
    /**
     * Return temporary selected minute (const MINUTES)
     * @return {?}
     */
    get minute() {
        return this._minute;
    }
    /**
     * Return temporary selected period (am/pm)
     * @return {?}
     */
    get period() {
        return this._period;
    }
    /**
     * Hide Confirm and Cancel buttons
     * @return {?}
     */
    get hideButtons() {
        return this._hideButtons;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set hideButtons(value) {
        this._hideButtons = coerceBooleanProperty(value);
    }
    /**
     * Return timer option class to create line between the middle of the clock and
     * the option
     * @return {?}
     */
    getSelectedClass() {
        let /** @type {?} */ name = 'selected-index-';
        if (this.focus === 'hour') {
            name += HOURS.findIndex(h => h === this.hour);
        }
        else {
            name += MINUTES.findIndex(m => m === this.minute);
        }
        return name;
    }
    /**
     * Select option from the clock.
     * @param {?} value MccTimerPickerHour | MccTimerPickerMinute
     * @return {?}
     */
    select(value) {
        if (this.focus === 'hour') {
            this._hour = /** @type {?} */ (value);
            this.focus = 'min';
        }
        else {
            this._minute = /** @type {?} */ (value);
        }
        // if buttons are hidden, emit new event when value is changed
        if (this._hideButtons) {
            this.confirmSelectedTime();
        }
    }
    /**
     * Change period of the clock
     * @param {?} period MccTimerPickerPeriod
     * @return {?}
     */
    changePeriod(period) {
        this._period = period;
        // if buttons are hidden, emit new event when value is changed
        if (this._hideButtons) {
            this.confirmSelectedTime();
        }
    }
    /**
     * Update selected color, close the panel and notify the user
     * @return {?}
     */
    backdropClick() {
        this.confirmSelectedTime();
        this._isOpen = false;
    }
    /**
     * Change values to last confirm select time
     * @return {?}
     */
    cancelSelection() {
        this._hour = this._selectedHour;
        this._minute = this._selectedMinute;
        this._period = this._selectedPeriod;
        this._isOpen = false;
    }
    /**
     * Set new values of time and emit new event with the formated timer
     * @return {?}
     */
    confirmSelectedTime() {
        this._selectedHour = this.hour;
        this._selectedMinute = this.minute;
        this._selectedPeriod = this.period;
        // format string to emit selected time
        let /** @type {?} */ formated;
        if (this.format === '12') {
            formated = `${this.hour}:${this.minute} ${this.period}`;
        }
        else {
            let /** @type {?} */ hour = this.hour;
            if (this.period === 'pm') {
                hour = `${parseInt(hour) + 12}`;
            }
            formated = `${hour}:${this.minute}`;
        }
        this.selected.emit(formated);
        // only close automatically if button aren't hidden
        if (!this._hideButtons) {
            this._isOpen = false;
        }
    }
}
MccTimerPickerComponent.decorators = [
    { type: Component, args: [{
                selector: 'mcc-timer-picker',
                template: `<!-- show component inside overlay -->
<ng-container *ngIf="connected">
  <ng-template cdkConnectedOverlay cdkConnectedOverlayHasBackdrop cdkConnectedOverlayBackdropClass="mcc-timer-picker-backdrop" [cdkConnectedOverlayOrigin]="trigger"
    [cdkConnectedOverlayOpen]="isOpen" (backdropClick)="backdropClick()">
    <ng-template [cdkPortalOutlet]="panelComponent"></ng-template>
  </ng-template>
</ng-container>
<ng-template *ngIf="!connected" [cdkPortalOutlet]="panelComponent"></ng-template>
<ng-template cdkPortal #panelComponent="cdkPortal">
  <div class="mcc-timer-picker-overlay mat-elevation-z6" role="dialog" aria-label="Timer picker">
    <div class="mcc-timer-picker-header mat-primary">
      <button mat-icon-button class="mcc-timer-picker-hours" [ngClass]="{ 'mcc-active': focus === 'hour' }" (click)="focus = 'hour'">
        {{ hour }}
      </button>
      <span class="mcc-timer-picker-separator"></span>
      <button mat-icon-button class="mcc-timer-picker-minutes" [ngClass]="{ 'mcc-active': focus === 'min' }" (click)="focus = 'min'">
        {{ minute }}
      </button>
      <div class="mcc-timer-picker-am-pm">
        <button mat-icon-button [ngClass]="{ 'mcc-active': period === 'am' }" (click)="changePeriod('am')">
          AM
        </button>
        <button mat-icon-button [ngClass]="{ 'mcc-active': period === 'pm' }" (click)="changePeriod('pm')">
          PM
        </button>
      </div>
    </div>
    <div class="mcc-timer-picker-content">
      <div class="mcc-timer-picker-clock">
        <div class="mcc-timer-picker-dot"></div>
        <div class="mcc-timer-picker-line" [ngClass]="getSelectedClass()"></div>
        <ng-container *ngFor="let option of clock$ | async;let i = index;">
          <button mat-icon-button class="mcc-timer-picker-option" [ngClass]="{ 'mcc-timer-picker-option-selected': focus === 'hour' && hour == option || focus === 'min' && minute === option }"
            [id]="'option-'+i" (click)="select(option)">
            {{ option }}
          </button>
        </ng-container>
      </div>
    </div>
    <div *ngIf="!hideButtons" class="mcc-timer-picker-actions" dir="rtl">
      <button mat-button role="button" aria-label="Confirm" (click)="confirmSelectedTime()">
        {{ btnConfirm }}
      </button>
      <button mat-button role="button" aria-label="Cancel" (click)="cancelSelection()">
        {{ btnCancel }}
      </button>
    </div>
  </div>
</ng-template>`,
                styles: [`@import url(https://fonts.googleapis.com/css?family=Roboto);.mcc-timer-picker-overlay{width:325px;height:480px}.mcc-timer-picker-header{display:-webkit-box;display:-ms-flexbox;display:flex;width:215px;height:65px;padding:15px 0 15px 110px;background:#00796b}.mcc-timer-picker-hours,.mcc-timer-picker-minutes{width:63px;height:63px;font-family:Roboto,sans-serif;font-size:50px;font-weight:400;color:rgba(255,255,255,.3)}.mcc-timer-picker-hours.mcc-standalone,.mcc-timer-picker-minutes.mcc-standalone{text-align:right}.mcc-timer-picker-hours.mcc-active,.mcc-timer-picker-minutes.mcc-active{color:#fff}.mcc-timer-picker-separator{font-family:Roboto,sans-serif;font-size:43px;font-weight:400;color:rgba(255,255,255,.3);margin:5px 5px 0 0}.mcc-timer-picker-separator::after{content:':'}.mcc-timer-picker-am-pm{display:-webkit-box;display:-ms-flexbox;display:flex;-webkit-box-orient:vertical;-webkit-box-direction:normal;-ms-flex-direction:column;flex-direction:column;margin:10px}.mcc-timer-picker-am-pm button{width:25px;height:25px;line-height:25px;font-family:Roboto,sans-serif;font-size:12px;font-weight:400;color:rgba(255,255,255,.3)}.mcc-timer-picker-am-pm button.mcc-active{color:#fff}.mcc-timer-picker-content{width:inherit;height:310px;padding:15px 0;background-color:#fff}.mcc-timer-picker-clock{position:relative;width:300px;height:300px;margin:0 auto;background-color:#eee;border-radius:50%;text-align:center}.mcc-timer-picker-clock .mcc-timer-picker-dot{position:absolute;width:6px;height:6px;border-radius:50%;background-color:#00796b;top:147px;left:147px}.mcc-timer-picker-clock .mcc-timer-picker-line{position:absolute;width:115px;height:2px;background-color:#00796b}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-0{top:92px;left:92px;-webkit-transform:rotateZ(90deg);transform:rotateZ(90deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-1{top:101px;left:58px;-webkit-transform:rotateZ(414deg);transform:rotateZ(414deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-2{top:100px;left:128px;-webkit-transform:rotateZ(126deg);transform:rotateZ(126deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-3{top:123px;left:39px;-webkit-transform:rotateZ(386deg);transform:rotateZ(386deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-4{top:124px;left:142px;-webkit-transform:rotateZ(153deg);transform:rotateZ(153deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-5{top:149px;left:34px;-webkit-transform:rotateZ(360deg);transform:rotateZ(360deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-6{top:149px;left:147px;-webkit-transform:rotateZ(180deg);transform:rotateZ(180deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-7{top:177px;left:42px;-webkit-transform:rotateZ(331deg);transform:rotateZ(331deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-8{top:178px;left:142px;-webkit-transform:rotateZ(210deg);transform:rotateZ(210deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-9{top:196px;left:58px;-webkit-transform:rotateZ(306deg);transform:rotateZ(306deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-10{top:195px;left:128px;-webkit-transform:rotateZ(233deg);transform:rotateZ(233deg)}.mcc-timer-picker-clock .mcc-timer-picker-line.selected-index-11{top:208px;left:92px;-webkit-transform:rotateZ(270deg);transform:rotateZ(270deg)}.mcc-timer-picker-clock .mcc-timer-picker-option{position:absolute}.mcc-timer-picker-clock .mcc-timer-picker-option.mcc-timer-picker-option-selected{color:#fff;background-color:#00796b}.mcc-timer-picker-clock .mcc-timer-picker-option#option-0{top:0;left:130px}.mcc-timer-picker-clock .mcc-timer-picker-option#option-1{top:25px;left:55px}.mcc-timer-picker-clock .mcc-timer-picker-option#option-2{top:25px;right:55px}.mcc-timer-picker-clock .mcc-timer-picker-option#option-3{top:70px;left:15px}.mcc-timer-picker-clock .mcc-timer-picker-option#option-4{top:70px;right:15px}.mcc-timer-picker-clock .mcc-timer-picker-option#option-5{top:130px;left:0}.mcc-timer-picker-clock .mcc-timer-picker-option#option-6{top:130px;right:0}.mcc-timer-picker-clock .mcc-timer-picker-option#option-7{bottom:65px;left:15px}.mcc-timer-picker-clock .mcc-timer-picker-option#option-8{bottom:65px;right:15px}.mcc-timer-picker-clock .mcc-timer-picker-option#option-9{bottom:25px;left:55px}.mcc-timer-picker-clock .mcc-timer-picker-option#option-10{bottom:25px;right:55px}.mcc-timer-picker-clock .mcc-timer-picker-option#option-11{bottom:0;left:130px}.mcc-timer-picker-actions{display:-webkit-box;display:-ms-flexbox;display:flex;padding:4px;background-color:#fff}.mcc-timer-picker-actions button{color:#100214;text-transform:uppercase;font-family:Roboto,sans-serif;font-size:14px;font-weight:400}`],
                preserveWhitespaces: false,
                changeDetection: ChangeDetectionStrategy.OnPush,
            },] },
];
/** @nocollapse */
MccTimerPickerComponent.ctorParameters = () => [];
MccTimerPickerComponent.propDecorators = {
    "hideButtons": [{ type: Input },],
    "format": [{ type: Input, args: ['mccTimerPickerFormat',] },],
    "btnCancel": [{ type: Input },],
    "btnConfirm": [{ type: Input },],
    "selected": [{ type: Output },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccTimerPickerOriginDirective {
    /**
     * Reference to the element on which the directive is applied.
     * @param {?} elementRef
     * @param {?} renderer
     */
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
        /**
         * Emit changes from the origin
         */
        this.change = new BehaviorSubject('');
        /**
         * Emit changes from the origin
         */
        this.hasFocus = new BehaviorSubject(false);
        // listen focus
        renderer.listen(elementRef.nativeElement, 'focus', () => this.hasFocus.next(true));
    }
    /**
     * This method will be called by the forms API to write to the view when
     * programmatic (model -> view) changes are requested.
     * @param {?} time
     * @return {?}
     */
    writeValue(time) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'value', time);
        this.change.next(time);
        if (this.propagateChanges) {
            this.propagateChanges(time);
        }
    }
    /**
     * This method will be called by the time picker
     * @param {?} time
     * @return {?}
     */
    writeValueFromTimerPicker(time) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'value', time);
        this.propagateChanges(time);
    }
    /**
     * This method will be called from origin whe key is up
     * @param {?} time
     * @return {?}
     */
    writeValueFromKeyup(time) {
        this.change.next(time);
        this.propagateChanges(time);
    }
    /**
     * This is called by the forms API on initialization so it can update the
     * form model when values propagate from the view (view -> model).
     * @param {?} fn any
     * @return {?}
     */
    registerOnChange(fn) {
        this.propagateChanges = fn;
    }
    /**
     * This is called by the forms API on initialization so it can update the form model on blur
     * @param {?} fn any
     * @return {?}
     */
    registerOnTouched(fn) { }
    /**
     * called by the forms API when the control status changes to or from "DISABLED"
     * @param {?} isDisabled boolean
     * @return {?}
     */
    setDisabledState(isDisabled) {
        this.renderer.setProperty(this.elementRef.nativeElement, 'disabled', isDisabled);
    }
}
MccTimerPickerOriginDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mccTimerPickerOrigin], [mcc-timer-picker-origin]',
                exportAs: 'mccTimerPickerOrigin',
                providers: [
                    {
                        provide: NG_VALUE_ACCESSOR,
                        useExisting: forwardRef(() => MccTimerPickerOriginDirective),
                        multi: true,
                    },
                ],
            },] },
];
/** @nocollapse */
MccTimerPickerOriginDirective.ctorParameters = () => [
    { type: ElementRef, },
    { type: Renderer2, },
];
MccTimerPickerOriginDirective.propDecorators = {
    "change": [{ type: Output },],
    "hasFocus": [{ type: Output },],
};
class MccConnectedTimerPickerDirective {
    /**
     * @param {?} timerPicker
     * @param {?} changeDetectorRef
     */
    constructor(timerPicker, changeDetectorRef) {
        this.timerPicker = timerPicker;
        this.changeDetectorRef = changeDetectorRef;
        this.timerPicker.connected = true;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (!this._timerPickerSub) {
            this.timerPicker.trigger = this.origin;
            this._attachTimerPicker();
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._originFocus && !this._originFocus.closed) {
            this._originFocus.unsubscribe();
        }
        if (this._timerPickerSub && !this._timerPickerSub.closed) {
            this._timerPickerSub.unsubscribe();
        }
    }
    /**
     * Attach the timer picker to origin element (input)
     * @return {?}
     */
    _attachTimerPicker() {
        this._originFocus = this.origin.hasFocus.subscribe(focused => {
            this.timerPicker.focus = 'hour';
            this.timerPicker.isOpen = focused;
            this.changeDetectorRef.detectChanges();
        });
        this._timerPickerSub = this.timerPicker.selected.subscribe(value => this.origin.writeValueFromTimerPicker(value));
    }
}
MccConnectedTimerPickerDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mccConnectedTimerPicker], [mcc-connected-timer-picker]',
                exportAs: 'mccConnectedTimerPicker',
            },] },
];
/** @nocollapse */
MccConnectedTimerPickerDirective.ctorParameters = () => [
    { type: MccTimerPickerComponent, },
    { type: ChangeDetectorRef, },
];
MccConnectedTimerPickerDirective.propDecorators = {
    "origin": [{ type: Input, args: ['mccConnectedTimerPickerOrigin',] },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccTimerPickerModule {
}
MccTimerPickerModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, PortalModule, OverlayModule, MatButtonModule],
                declarations: [
                    MccTimerPickerComponent,
                    MccTimerPickerOriginDirective,
                    MccConnectedTimerPickerDirective,
                ],
                exports: [
                    MccTimerPickerComponent,
                    MccTimerPickerOriginDirective,
                    MccConnectedTimerPickerDirective,
                ],
            },] },
];
/** @nocollapse */
MccTimerPickerModule.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
// scrollspy animations
const SCROLLSPY_ANIMATION_SMOOTH = 'smooth';


/**
 * Scrollspy group
 * @record
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccScrollspyService {
    /**
     * @param {?} window
     */
    constructor(window) {
        this.window = window;
        /**
         * When scroll is from click event, change this attr to true
         * So scroll event obeservable doesn't emit any update
         */
        this._fromClick = false;
        /**
         * List of scrollspy group
         */
        this.data = [];
        // listen to scroll event
        this._scrollSub = fromEvent(window, 'scroll')
            .pipe(debounceTime(50), withLatestFrom(() => window.scrollY))
            .subscribe(position => {
            if (!this._fromClick) {
                this._updateFocused(position);
            }
            this._fromClick = false;
        });
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        if (this._scrollSub && !this._scrollSub.closed) {
            this._scrollSub.unsubscribe();
        }
    }
    /**
     * Update information about wich element is on focus
     * @param {?} position number
     * @return {?}
     */
    _updateFocused(position) {
        this.data.forEach(group => {
            const /** @type {?} */ items = [];
            group.items.getValue().forEach((item, index) => {
                item.focus = false;
                if (item.top <= position) {
                    if (items[index - 1]) {
                        items[index - 1].focus = false;
                    }
                    item.focus = true;
                }
                items.push(item);
            });
            group.items.next(items);
        });
    }
    /**
     * Create new group of items
     * @param {?} name string
     * @param {?=} items MccScrollspyItemDirective[]
     * @param {?=} animation ScrollBehavior
     * @return {?}
     */
    create(name, items, animation) {
        let /** @type {?} */ group = this.data.find(group => group.name === name);
        if (!group) {
            group = {
                name: name,
                animation: animation || SCROLLSPY_ANIMATION_SMOOTH,
                items: new BehaviorSubject(items || []),
            };
            this.data.push(group);
        }
        else {
            group.items.next(items || []);
        }
        return group;
    }
    /**
     * Return observable of the group
     * @param {?} name string
     * @return {?}
     */
    group(name) {
        let /** @type {?} */ group = this.data.find(g => g.name === name);
        if (!group) {
            group = this.create(name);
        }
        return group.items.asObservable();
    }
    /**
     * Scroll to one of the items
     * @param {?} name string
     * @param {?} id string
     * @return {?}
     */
    scrollTo(name, id) {
        const /** @type {?} */ group = this.data.find(group => group.name === name);
        group.items.getValue().forEach(item => {
            item.focus = false;
            if (item.id === id) {
                this._fromClick = true;
                this._updateFocused(item.top);
                window.scrollTo({ top: item.top, behavior: group.animation });
            }
        });
    }
}
MccScrollspyService.decorators = [
    { type: Injectable },
];
/** @nocollapse */
MccScrollspyService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: ['Window',] },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
class MccScrollspyItemDirective {
    /**
     * @param {?} elementRef
     * @param {?} renderer
     */
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    /**
     * Hold the element id, if element doesn't have id
     * the method will create one
     * @param {?} id
     * @return {?}
     */
    set id(id) {
        if (!id) {
            id = this._createId();
            this.renderer.setProperty(this.elementRef.nativeElement, 'id', id);
        }
        this._id = id;
    }
    /**
     * @return {?}
     */
    get id() {
        return this._id;
    }
    /**
     * Element distance of the top
     * @return {?}
     */
    get top() {
        return this.elementRef.nativeElement.offsetTop;
    }
    /**
     * Element is focused
     * @param {?} focused
     * @return {?}
     */
    set focus(focused) {
        this._focused = coerceBooleanProperty(focused);
    }
    /**
     * @return {?}
     */
    get focus() {
        return this._focused;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        if (!this.label) {
            this.label = this.elementRef.nativeElement.textContent;
        }
        this.id = this.elementRef.nativeElement.id;
    }
    /**
     * Create an ID for the element
     * @return {?}
     */
    _createId() {
        let /** @type {?} */ tmpID = this.label.toLowerCase().replace(/[ ]+/gi, '_');
        return `mcc_scrollspy_${tmpID}`;
    }
}
MccScrollspyItemDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mccScrollspyItem], [mcc-scrollspy-item]',
                exportAs: 'mccScrollspyItem',
            },] },
];
/** @nocollapse */
MccScrollspyItemDirective.ctorParameters = () => [
    { type: ElementRef, },
    { type: Renderer2, },
];
MccScrollspyItemDirective.propDecorators = {
    "focus": [{ type: Input, args: ['focus',] },],
    "label": [{ type: Input },],
};
class MccScrollspyGroupDirective {
    /**
     * @param {?} mccScrollspyService
     */
    constructor(mccScrollspyService) {
        this.mccScrollspyService = mccScrollspyService;
    }
    /**
     * Name of the scrollspy group
     * @param {?} name
     * @return {?}
     */
    set name(name) {
        this._name = name;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        // add the group items
        const /** @type {?} */ items = this.items.map(item => item);
        this.mccScrollspyService.create(this._name, items);
    }
}
MccScrollspyGroupDirective.decorators = [
    { type: Directive, args: [{
                selector: '[mccScrollspyGroup], [mcc-scrollspy-group]',
                exportAs: 'mccScrollspyGroup',
            },] },
];
/** @nocollapse */
MccScrollspyGroupDirective.ctorParameters = () => [
    { type: MccScrollspyService, },
];
MccScrollspyGroupDirective.propDecorators = {
    "items": [{ type: ContentChildren, args: [MccScrollspyItemDirective,] },],
    "name": [{ type: Input, args: ['mccScrollspyGroup',] },],
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
const 0 = window;
class MccScrollspyModule {
}
MccScrollspyModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, ScrollDispatchModule],
                providers: [MccScrollspyService, { provide: 'Window', useValue: 0 }],
                declarations: [MccScrollspyGroupDirective, MccScrollspyItemDirective],
                exports: [MccScrollspyGroupDirective, MccScrollspyItemDirective],
            },] },
];
/** @nocollapse */
MccScrollspyModule.ctorParameters = () => [];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { MccColorPickerModule, MccColorPickerService, EMPTY_COLOR, MccTimerPickerModule, MccScrollspyModule, MccScrollspyService, MccScrollspyItemDirective, MccColorPickerCollectionComponent as c, MccColorPickerSelectorComponent as g, MccColorPickerComponent as b, MccColorPickerOptionDirective as d, MccColorPickerOriginDirective as e, MccConnectedColorPickerDirective as f, MccScrollspyGroupDirective as a, MccTimerPickerComponent as h, MccConnectedTimerPickerDirective as j, MccTimerPickerOriginDirective as i };
